src/main/java/com/iota/iri/hash/Kerl.java

```diff
@@ -6,142 +6,174 @@
 import java.math.BigInteger;
 import java.security.DigestException;
 import java.util.Arrays;
+import java.util.stream.IntStream;
+
+public final class Kerl implements Sponge {
 
-public class Kerl implements Sponge {
     public static final int BIT_HASH_LENGTH = 384;
     public static final int BYTE_HASH_LENGTH = BIT_HASH_LENGTH / 8;
 
-    private byte[] byte_state;
-    private int[] trit_state;
+    // radix constant conversion for fast math
+    public static final BigInteger RADIX = BigInteger.valueOf(Converter.RADIX);
+    private static final BigInteger[] RADIX_POWERS = IntStream.range(0, Sponge.HASH_LENGTH + 1).mapToObj(RADIX::pow).toArray(BigInteger[]::new);
+
+    private final byte[] byte_state;
+    private final int[] trit_state;
     private final Keccak.Digest384 keccak;
 
     protected Kerl() {
-
         this.keccak = new Keccak.Digest384();
         this.byte_state = new byte[BYTE_HASH_LENGTH];
         this.trit_state = new int[Sponge.HASH_LENGTH];
     }
 
     @Override
     public void reset() {
-
         this.keccak.reset();
     }
 
     @Override
-    public void absorb(final int[] trits, int offset, int length) {
-
+    public void absorb(final int[] trits, final int offset, final int length) {
         if (length % 243 != 0) {
             throw new RuntimeException("Illegal length: " + length);
         }
-
-        do {
+        for (int pos = offset; pos < offset + length; pos += HASH_LENGTH) {
             //copy trits[offset:offset+length]
-            System.arraycopy(trits, offset, trit_state, 0, HASH_LENGTH);
+            System.arraycopy(trits, pos, trit_state, 0, HASH_LENGTH);
 
             //convert to bits
             trit_state[HASH_LENGTH - 1] = 0;
-            bytesFromBigInt(bigIntFromTrits(trit_state, 0, HASH_LENGTH), byte_state, 0);
+            bytesFromBigInt(bigIntFromTrits(trit_state, 0, HASH_LENGTH), byte_state);
 
             //run keccak
             keccak.update(byte_state);
-            offset += HASH_LENGTH;
-
-        } while ((length -= HASH_LENGTH) > 0);
+        }
     }
```

Comment:
I just thought it reads clearer with a traditional for loop than tracking both offset and length in the old version. Now the inputs are also final.